Python/syntax_test_python_strings.py

```diff
@@ -549,6 +549,45 @@
 bar = "}}" # Comment
 #      ^^ constant.character.escape
 
+# The following section contains unusual and legal or illegal format placeholders.
+# We don't actually want to match the syntax 100% of the time,
+# since we never know for sure if the string is used as a format string,
+# so some of these matches are incorrect because of implementation details.
+
+# Not format specs
+"{:{ }"  # unclosed
+# ^ - constant.other.placeholder
+'{{foo!r:4.2}'  # escaped opening
+# ^ - constant.other.placeholder
+'{{foo!r:4.2}}'  # escaped opening and closing
+# ^ - constant.other.placeholder
+'{foo!a:ran{dom}'  # unclosed
+# ^ - constant.other.placeholder
+'{foo!a:ran{dom}'  # unclosed
+# ^ - constant.other.placeholder
+
+# Invalid field names
+'{foo{d}}'
+# ^ - constant.other.placeholder
+'{foo.!a:d}'
+# ^ constant.other.placeholder
```

Comment:
That is correct-ish. Trying to format this raises `ValueError: Empty attribute in format string`, but in reality it's the same as typing `foo.` outside of a format string which we still highlight because of a) convenience for the highlighter and b) highlighting while typing.